haystack/core/pipeline/utils.py

```diff
@@ -24,6 +26,11 @@ def deepcopy_with_fallback(obj: Any, max_depth: Optional[int] = 100) -> Any:
     :param max_depth: The maximum depth to recurse during deep copying. If 0, returns the object as-is.
     :return: A deep copy of the object if successful; otherwise, the original object.
     """
+    # Components and Tools often contain objects that we do not want to deepcopy or are not deepcopyable
+    # (e.g. pytorch models, clients, etc.). In this case we return the object as-is.
+    if isinstance(obj, (Component, Tool, Toolset)):
+        return obj
```

Comment:
I'm wondering actually now that I think of it more if we should instead create custom `__deepcopy__` functions for each of these classes to just return itself so 
```python
def __deepcopy__(self):
    return self
```

Reply:
However, this could potentially be undesirable in cases users create custom Component, Tool or Toolsets that they would in fact like to deepcopy.